### DISCUSSION TITLE
ESPNow MQTT gateway

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#11894] [ESP32]
@glenn20 I have published some code [here](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/esp32_gateway/GATEWAY.md) and would welcome feedback. You'll notice I've taken your name in vain in [section 6.3](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/esp32_gateway/GATEWAY.md#63-case-where-wifi-channel-varies) - comments here would be appreciated.

For those unfamiliar with our discussion the tl;dr is that ESPx nodes can act as very simple synchronous MQTT clients, optionally designed for ultra low power consumption. Nodes communicate with an ESP32 gateway device which runs [asynchronous MQTT](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/README.md) to reliably communicate with an MQTT broker, which may be local or online.

Code works but needs a lot more testing, notably with nodes on ESP8266. Encryption is not yet supported.

I am particularly looking for feedback on the overall design.

**Comment by glenn20** [comment]
Hi Peter, Sorry for the slow response. Thanks for sharing - this is really useful. I'll look at testing your code this week.

**Comment by peterhinch** [comment]
I believe I've thought of a better way of handling subscriptions. I'll start work on this tomorrow so the API for publications to the gateway may well change, with the nodes acquiring a subscribe function.

The aim is to avoid the need for external devices knowing the ID of the target node; also to allow an external device to publish to a topic, with one or more nodes subscribing to that topic.

**Comment by peterhinch** [comment]
I've now pushed an update with a much cleaner API for node subscriptions. One improvement is that the only MAC address you need to know is that of the gateway.

**Comment by peterhinch** [comment]
@glenn20 To avoid noise in discussions I raised [this issue](https://github.com/peterhinch/micropython-mqtt/issues/116) which I'd be grateful if you could study when you get a chance.
